src/node_api.cc

```diff
@@ -46,26 +47,34 @@ void node_napi_env__::CallFinalizer(napi_finalize cb, void* data, void* hint) {
 
 template <bool enforceUncaughtExceptionPolicy>
 void node_napi_env__::CallFinalizer(napi_finalize cb, void* data, void* hint) {
-  if (destructing) {
-    // we can not defer finalizers when the environment is being destructed.
-    v8::HandleScope handle_scope(isolate);
-    v8::Context::Scope context_scope(context());
-    CallbackIntoModule<enforceUncaughtExceptionPolicy>(
-        [&](napi_env env) { cb(env, data, hint); });
-    return;
+  v8::HandleScope handle_scope(isolate);
+  v8::Context::Scope context_scope(context());
+  CallbackIntoModule<enforceUncaughtExceptionPolicy>(
+      [&](napi_env env) { cb(env, data, hint); });
+}
+
+void node_napi_env__::EnqueueFinalizer(v8impl::RefTracker* finalizer) {
+  napi_env__::EnqueueFinalizer(finalizer);
+  if (!finalization_scheduled) {
+    finalization_scheduled = true;
+    Ref();
+    node_env()->SetImmediate([this](node::Environment* node_env) {
```

Comment:
Capturing `this` is not safe: if the lambda is not executed, then the `env` instance is never released. It would be better to use the `EnvRefHolder` here which is being deleted in this PR from js_native_api_v8.h. This way the Unref() is always called when the lambda is released.
Also, when we call the `Unref()` below we must be prepared that it was the last `Unref()` that causes the `env` instance destruction. In such case the `DrainFinalizerQueue()` may be called against deleted object and cause issues.

Reply:
The callback of `Environment::SetImmediate` is deemed to be called. So we don't need additional setups for the condition that the callback is not invoked.